main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -190,82 +378,497 @@ class NetworkClient(configuration: xsbti.AppConfiguration, arguments: List[Strin
+    val stopped = new AtomicBoolean(false)
+    val lock = new Object
+    override final def run(): Unit = {
+      @tailrec def read(): Unit = {
+        inputStream.read match {
+          case -1 =>
+          case b =>
+            lock.synchronized(stdinBytes.offer(b))
+            if (attached.get()) drain()
+            if (!stopped.get()) read()
+        }
+      }
+      try Terminal.console.withRawSystemIn(read())
+      catch { case _: InterruptedException | _: ClosedChannelException => stopped.set(true) }
+    }
+
+    def drain(): Unit = lock.synchronized {
+      while (!stdinBytes.isEmpty) {
+        val byte = stdinBytes.poll()
+        sendNotification(systemIn, byte.toString)
+      }
+    }
+
+    override def close(): Unit = {
+      RawInputThread.this.interrupt()
+    }
+  }
+
+  // copied from Aggregation
+  private def timing(startTime: Long, endTime: Long): String = {
+    import java.text.DateFormat
+    val format = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.MEDIUM)
+    val nowString = format.format(new java.util.Date(endTime))
+    val total = (endTime - startTime + 500) / 1000
+    val totalString = s"$total s" +
+      (if (total <= 60) ""
+       else {
+         val maybeHours = total / 3600 match {
+           case 0 => ""
+           case h => f"$h%02d:"
+         }
+         val mins = f"${total % 3600 / 60}%02d"
+         val secs = f"${total % 60}%02d"
+         s" ($maybeHours$mins:$secs)"
+       })
+    s"Total time: $totalString, completed $nowString"
+  }
+  override def close(): Unit =
+    try {
+      running.set(false)
+      stdinBytes.offer(-1)
+      val mainThread = interactiveThread.getAndSet(null)
+      if (mainThread != null && mainThread != Thread.currentThread) mainThread.interrupt
+      connection.shutdown()
+      Option(inputThread.get).foreach(_.interrupt())
+    } catch {
+      case t: Throwable => t.printStackTrace(); throw t
+    }
+}
+class NetworkClient(
+    override val arguments: Arguments,
+    override val inputStream: InputStream,
+    override val errorStream: PrintStream,
+    override val printStream: PrintStream,
+) extends {
+  override val console: ConsoleInterface = {
+    val appender = ConsoleAppender("thin")
+    new ConsoleInterface {
+      override def appendLog(level: Level.Value, message: => String): Unit =
+        appender.appendLog(level, message)
+      override def success(msg: String): Unit = appender.success(msg)
+    }
+  }
+} with NetworkClientImpl {
+  def this(configuration: xsbti.AppConfiguration, arguments: Arguments) =
+    this(
+      arguments.withBaseDirectory(configuration.baseDirectory),
+      System.in,
+      System.err,
+      System.out,
+    )
+  override def mkSocket(file: File): (Socket, Option[String]) = ClientSocket.socket(file)
+  def this(configuration: xsbti.AppConfiguration, args: List[String]) =
+    this(
+      NetworkClient.parseArgs(args.toArray).withBaseDirectory(configuration.baseDirectory),
```

Comment:
Which one's meant to win `--sbt-base-directory` or `configuration.baseDirectory`?  It seems like `--sbt-base-directory` never wins, except for completions?  I feel like the base directory (or the `new File("")` fallback) should be an argument of `parseArgs`.

Reply:
I'll look into this. This was added way back when I had problems getting runtime initialization of graalvm classes to work. At the time `new File("")` was returning a build path so I added the `--sbt-base-directory` to address that. I fixed the `new File("")` issue but left in `--sbt-base-directory`. It probably should be removed. Just ftr, today I am cleaning up `NetworkClient.scala` and `NetworkChannel.scala` so these files will probably look somewhat different by the end of the day.

Reply:
Same here, spent a few hours learning bits of how it works together.

Let's just say I wish scalafmt added newline between multi-line methods!! o_O

Reply:
Urelated comment to this PR.

But @dwijnand there is such a config btw, I guess the problem will be when to apply a new scalafmt config to the codebase 😅 

https://scalameta.org/scalafmt/docs/configuration.html#newlinesalwaysbeforemultilinedef

Reply:
Thanks, @lorandszakacs!  Looks like I want: 

```
newlines.topLevelStatements = [before,after]
```